### DISCUSSION TITLE
Added: LDM Notes for April 22 and April 24, 2019

### DISCUSSION THREAD

**Discussion by agocke** [dotnet/csharplang#2471] [General]
- https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-04-22.md

- https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-04-24.md

**Comment by orthoxerox** [comment]
Wouldn't an unconstrained `T?` imply that the method returns a `Nullable<T>` for structs?

**Comment by gafter** [comment]
@orthoxerox No, and that the main thing that makes us concerned about it.

**Comment by qrli** [comment]
Ah... I finally understand the concerns. That's tricky.
Having been a R# nullability attribute user, I'm pretty fine with the attribute annotated approach. It looks already more powerful than R#'s.

**Comment by Joe4evr** [comment]
> However, we think `MaybeNull` may be common or impactful enough to consider syntax, like allowing
`T?` on an unconstrained type parameter or `T : object?`, to express `MaybeNull`.

I'm not a huge fan of using `T?` exactly as-is to express `MaybeNull` for unconstrained `T` return values, and would rather see it differently or not at all (meaning: just an unannotated `T` + attribute on the method). It does seem like something that can easily be added on later without breaking, so maybe be a bit more pragmatic and don't let this one get in the way of shipping 8.0.

Looking through the list of "problematic" signatures, I think some of them don't need to be on there:
> - WeakReference<T>.TryGetTarget(out T target)

Isn't `WeakReference` constrained to `class` anyway? That would make `out T? target` an obvious solution. Other `out` parameters could go with the `[MaybeNull]` approach.
> Interfaces where T should be nullable on some methods but non-nullable on others, e.g.

Probably also a good candidate for `[MaybeNull]`.
> Methods where whether the return value can be null is impacted by whether an argument is null

I thought that this is what `[NullInNullOut]` is for. (dotnet/roslyn#26761)
> Ref arguments that will be non-null upon return

And `[NotNulled]` is for this. (same as above)
> Methods that accept an delegate<object?> and object?, but where the argument to Action<object?> will be non-null iff the object? is non-null

On-the-spot idea: `[NullablePassthrough]` for the `object?`, indicating that the delegate's argument takes on the nullability of the annotated parameter.
```cs
// example using one of System.Threading.Timer's ctors:
public Timer(TimerCallback callback, [NullablePassthrough(0, 0)] object? state, TimeSpan dueTime, TimeSpan period) { }
```
The `(0, 0)` arguments are like a navigation to which method parameter is the target delegate is and which delegate parameter is the target argument.

Okay, so there's probably a whole host of other things that would have to be considered for that last one, but an idea's an idea.

**Comment by yaakov-h** [comment]
> Isn't `WeakReference` constrained to `class` anyway?

ooh, good catch. But wouldn't `TryGetTarget` be a use case for `NotNullWhenTrueAttribute` anyway?

**Comment by agocke** [comment]
@Joe4evr Good points. One note: those "problematic signatures" are just things that can't be addressed without attributes or additional syntax. So it includes stuff that could be addressed using attributes (I don't know about WeakReference, that's good to check).
